flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java

```diff
@@ -865,9 +900,20 @@ public JobExecutionResult execute(String jobName) throws Exception {
 			"Cannot find compatible factory for specified execution.target (=%s)",
 			configuration.get(DeploymentOptions.TARGET));
 
-		return executorFactory
+		CompletableFuture<JobClient> jobClientFuture = executorFactory
 			.getExecutor(configuration)
 			.execute(plan, configuration);
+
+		jobClientFuture.whenCompleteAsync((jobClient, throwable) -> {
```

Comment:
I think we should return the result of `whenCompleteAsync`, this way, the returned future only returns when the job listener was invoked and has returned. This ensures that the job listener can still create a job client if necessary. Otherwise we have a race condition if the consumer of the returned future immediately closes the client. I think.

Reply:
With implementation detail, the resource can be re-created by configuration after closed. The concern we handle it asynchronously is because `JobListener`'s callbacks can be expensive. We can do as you said with mentioning DO NOT invoke expensive calls in callbacks

Reply:
Ah, both `JobClient` and `ClusterClient` can be duplicated after they have been closed. Maybe we should add a test to guard that behaviour then.